### PR TITLE
Added Example for CustomListener with StatusMask [9042]

### PR DESCRIPTION
**Pull request by imontesino**

### CODE REVIEW

File: examples/C++/DDS/CustomListenerExample/CustomListenerPublisher.cpp [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        return false;
+    }
+
+    // CREATE THE WRITER
+    StatusMask none_mask = StatusMask::none();
+    StatusMask all_mask = StatusMask::all();
+    dw_mask_ = use_dw ? all_mask : none_mask;
+    
+    writer_ = publisher_->create_datawriter(topic_, DATAWRITER_QOS_DEFAULT, &dw_listener_, dw_mask_);
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/CustomListenerPublisher.cpp:78]: We can avoid the local variables and the member `dw_mask_` altogether.
```suggestion
    writer_ = publisher_->create_datawriter(topic_, DATAWRITER_QOS_DEFAULT, &dw_listener_,
            use_dw ? StatusMask::all() : StatusMask::none());
```

File: examples/C++/DDS/CustomListenerExample/CustomListenerPublisher.h [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    eprosima::fastdds::dds::DomainParticipant* participant_;
+
+    eprosima::fastdds::dds::Publisher* publisher_;
+
+    eprosima::fastdds::dds::Topic* topic_;
+
+    eprosima::fastdds::dds::DataWriter* writer_;
+
+    eprosima::fastdds::dds::StatusMask dw_mask_;
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/CustomListenerPublisher.h:62]: We can remove this with the previous change

File: examples/C++/DDS/CustomListenerExample/CustomListenerSubscriber.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        return false;
+    }
+
+    // CREATE THE READER
+    StatusMask none_mask = StatusMask::none();
+    StatusMask all_mask = StatusMask::all();
+    dr_mask_ = use_dr ? all_mask : none_mask;
+    DataReaderQos rqos = DATAREADER_QOS_DEFAULT;
+    rqos.reliability().kind = RELIABLE_RELIABILITY_QOS;
+    reader_ = subscriber_->create_datareader(topic_, rqos, &dr_listener_, dr_mask_);
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/CustomListenerSubscriber.cpp:83]: Same as with the publisher
```suggestion
    DataReaderQos rqos = DATAREADER_QOS_DEFAULT;
    rqos.reliability().kind = RELIABLE_RELIABILITY_QOS;
    reader_ = subscriber_->create_datareader(topic_, rqos, &dr_listener_,
             use_dr ? StatusMask::all() : StatusMask::none());
```

File: examples/C++/DDS/CustomListenerExample/CustomListenerSubscriber.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    eprosima::fastdds::dds::Subscriber* subscriber_;
+
+    eprosima::fastdds::dds::Topic* topic_;
+
+    eprosima::fastdds::dds::DataReader* reader_;
+
+    eprosima::fastdds::dds::TypeSupport type_;
+
+    eprosima::fastdds::dds::StatusMask dr_mask_;
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/CustomListenerSubscriber.h:62]: Can be removed with the previous change

File: examples/C++/DDS/CustomListenerExample/README.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+The above will launch publisher and subscriber with the listener of the bottom most entities activated.
+
+## Application behaviour
+
+In this application, the publisher sends a number of samples to the subscriber. In the publisher application both the
+DomainParticipant and the DataWriter have listeners with the `on_publication_matched` callback implemented defined.
+When the ENABLE_LOWESTLISTENER argument is set to false the DataWriterListener is given a `StatusMask::none()` which
+disables all of its callbacks.
+These callbacks are therefor intercepted by the next implemented callback. In this case the DomainParticipant.
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/README.md:25]:
```suggestion
These callbacks are therefore intercepted by the next implemented callback. In this case the DomainParticipant.
```

File: examples/C++/DDS/CustomListenerExample/CustomListener_main.cpp [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    bool use_d;
+
+    if (argc > 2)
+    {
+        if (strcmp(argv[1], "publisher") == 0)
+        {
+            type = 1;
+            if (argc >= 4)
+            {
+                count = atoi(argv[2]);
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/CustomListener_main.cpp]:
```suggestion
                count = atoi(argv[3]);
```

File: examples/C++/DDS/CustomListenerExample/CustomListener_main.cpp [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                if (argc == 5)
+                {
+                    sleep = atoi(argv[3]);
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/CustomListener_main.cpp]:
```suggestion
                    sleep = atoi(argv[4]);
```

File: examples/C++/DDS/CustomListenerExample/CustomListenerPublisher.cpp [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        std::cout << "Publisher running " << samples << " samples." << std::endl;
+    }
+    thread.join();
+}
+
+bool CustomListenerPublisher::publish(
+        bool waitForListener)
+{
+    if (dw_listener_.firstConnected_ || !waitForListener || dw_listener_.matched_ > 0 ||
+        p_listener_.firstConnected_  || p_listener_.matched_ > 0)
```
Comment (IkerLuengo) [inline examples/C++/DDS/CustomListenerExample/CustomListenerPublisher.cpp:163]: I know this comes from the basic example, but `XX_listener_.matched_ > 0` is redundant with `XX_listener_.firstConnected_`, as the latter is set to true as soon as there is at least one match, and is never reverted back even if the number of matches comes to zero.
```suggestion
    if (!waitForListener || dw_listener_.firstConnected_ || p_listener_.firstConnected_)
```

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/2477/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2290/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/2481/)

**Pull request merged**